naga/src/common/diagnostic_display.rs

```diff
@@ -49,6 +49,17 @@ use core::fmt;
 /// indication will provoke a compile-time error.
 pub struct DiagnosticDisplay<T>(pub T);
 
+impl fmt::Display for DiagnosticDisplay<(&TypeResolution, GlobalCtx<'_>)> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let (resolution, ctx) = self.0;
+
+        match *resolution {
+            TypeResolution::Handle(handle) => DiagnosticDisplay((handle, ctx)).fmt(f),
+            TypeResolution::Value(ref inner) => DiagnosticDisplay((inner, ctx)).fmt(f),
+        }
+    }
+}
+
 impl fmt::Display for DiagnosticDisplay<(Handle<Type>, GlobalCtx<'_>)> {
```

Comment:
question: `DiagnosticDisplay` with a `Handle<Type>` has a footgun in that it can't handle `struct`s. Should we try to remove this `Display` implementation, and force other use sites to migrate?

Reply:
This was a serious problem. I've greatly expanded the PR.